[Salesforce] Add pagination support

kush-elastic opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
- Cleanup
- Docs
-->

## What does this PR do?
In the current Salesforce Module in filebeat, we don’t have pagination support for all four file-sets (login-rest, logout-rest, apex-rest,  setupaudittrail-rest). So if in any case, we have a response body with more than 2000 records, the next page is never called and the data will not be fetched.

Solution:
You can request the next batch of records using the provided URL with your instance and session information, and repeat until all records have been retrieved. These requests use `nextRecordsUrl` and don’t include any parameters. The final batch of records doesn’t have a `nextRecordsUrl` field and return `true` value for field `done`

Ref Link for Issue:  https://blog.bessereau.eu/assets/pdfs/api_rest.pdf
<!-- Mandatory
Explain here the changes you made on the PR. Please explain the WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
-->

## Why is it important?

<!-- Mandatory
Explain here the WHY, or the rationale/motivation for the changes.
-->

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [X] My code follows the style guidelines of this project
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have made corresponding change to the default configuration files
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

<!-- ## Author's Checklist

Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved

- [ ]
-->
<!-- ## How to test this PR locally

Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Requires #123
- Superseds #123
-->
- Relates #34057 
 
<!-- ## Use cases

Recommended
Explain here the different behaviors that this PR introduces or modifies in this project, user roles, environment configuration, etc.

If you are familiar with Gherkin test scenarios, we recommend its usage: https://cucumber.io/docs/gherkin/reference/
-->

<!-- ## Screenshots

Optional
Add here screenshots about how the project will be changed after the PR is applied. They could be related to web pages, terminal, etc, or any other image you consider important to be shared with the team.
-->

<!-- ## Logs

Recommended
Paste here output logs discovered while creating this PR, such as stack traces or integration logs, or any other output you consider important to be shared with the team.
-->

elasticmachine commented:
## :green_heart: Build Succeeded


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-34065/detail/PR-34065/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-34065/detail/PR-34065/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-34065/detail/PR-34065/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-34065/detail/PR-34065/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://beats_34065.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2022-12-19T05:06:58.917Z&rangeTo=2022-12-19T05:26:58.917Z&transactionName=BUILD+Beats%2Fbeats%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=9f48287780e89647268ad16443468f59&transactionId=174819a88bc9a596)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-12-19T05:16:58.917+0000


* Duration: 80 min 10 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 2552  |
| Skipped      | 167 |
| Total        | 2719   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














## :green_heart: Flaky test report
Tests succeeded.



## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate the packages and run the E2E tests.


- `/beats-tester` : Run the installation tests with beats-tester.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

agithomas commented:
CreatedDate is changed to LogDate in the query.  Which one is most appropriate? Why ?
Is this change fixing any bug?

Kindly complete the internal code review.

kush-elastic commented:
> CreatedDate is changed to LogDate in the query. Which one is most appropriate? Why ? Is this change fixing any bug?
> 

Currently, in the httpjson we are not able to use the pagination + split + chain call's , we had a similar Issue(https://github.com/elastic/beats/issues/33213) previously and we are in a chat with @ShourieG  who has worked on this issue previously(https://github.com/elastic/beats/pull/33664), He is currently looking into the issue, we have provided the required configurations to dig deep into this issue.

Issue
We are not getting the response for the second page.

Solution 
-> We removed the split functionality which was used to set cursor value of (CreatedDate) from the first response so that we are able to receive the response from the second page.

-> We changed the cursor value to more reliable data that we receive in the second call(chain call) TIMESTAMP_DERIVED which is related to the value between (LogDate  and LogDate + 1h) and has no relation with CreatedDate.

-> We have tried data collection without split functionality it's working fine as it was used for just setting the cursor value which is now changed to TIMESTAMP_DERIVED. Also, we have updated the query to LogDate as it is more related to TIMESTAMP_DERIVED.

Also, we have used the similar logic in our salesforce integration as well.
JFYI:

- LogDate tracks usage activity of a log file for a 24-hour period, from 12:00 a.m. to 11:59 p.m.
- CreatedDate tracks when the log file was generated.

> Kindly complete the internal code review.

This PR is internally reviewed but as discussed [here](https://github.com/elastic/integrations/pull/4866#issuecomment-1360915712) we don't have any other account as of now to provide approval, so consider this PR approved, Harnish can comment the same.

harnish-elastic commented:
LGTM! Approved.

agithomas commented on x-pack/filebeat/module/salesforce/logout-rest/config/logout-rest.yml:
```diff
@@ -29,18 +29,21 @@ request.proxy_url: {{ .proxy_url }}
-      default: "SELECT Id,CreatedDate,LogDate,LogFile FROM EventLogFile WHERE Interval = 'Hourly' AND EventType = 'Logout' ORDER BY CreatedDate ASC NULLS FIRST"
-response.split:
-  target: body.records
+      value: "SELECT Id,CreatedDate,LogDate,LogFile FROM EventLogFile WHERE Interval = 'Hourly' AND EventType = 'Logout' AND LogDate > [[.cursor.last_published_logout]] ORDER BY CreatedDate ASC NULLS FIRST"
```
The orderby clause must be `LogDate` . 
Please make the similar correction in 

https://github.com/elastic/integrations/blob/main/packages/salesforce/data_stream/logout_rest/agent/stream/httpjson.yml.hbs

agithomas reviewed (approved):
Approving. 

We need to see the outcome of testcases run to resolve the SDH issue.  According to the outcome,  the queries may need to be modified.
